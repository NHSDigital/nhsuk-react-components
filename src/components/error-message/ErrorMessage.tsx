import React, { HTMLProps, useContext } from 'react';

import classNames from 'classnames';
import FormGroupContext from '../formgroup/FormGroupContext';

export interface ErrorMessageProps extends HTMLProps<HTMLSpanElement> {
  visuallyHiddenText?: false | string;
}

const ErrorMessage: React.FC<ErrorMessageProps> = props => <BaseErrorMessage {...props} />;

const BaseErrorMessage: React.FC<ErrorMessageProps> = props => {
  const { id, className, visuallyHiddenText, children, ...rest } = props;
  const { inputID } = useContext(FormGroupContext);
  const errorMessageId = props.id || inputID || undefined;
  return (
    <span id={errorMessageId} className={classNames('nhsuk-error-message', className)} {...rest}>
      {visuallyHiddenText !== false ? (
        <span className="nhsuk-u-visually-hidden">{visuallyHiddenText}</span>
      ) : null}
      {children}
    </span>
  );
};

ErrorMessage.defaultProps = {
  visuallyHiddenText: 'Error: ',
  role: 'alert',
};

export default ErrorMessage;
