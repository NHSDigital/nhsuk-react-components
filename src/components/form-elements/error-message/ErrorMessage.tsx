import React, { HTMLProps } from 'react';
import classNames from 'classnames';

export interface ErrorMessageProps extends HTMLProps<HTMLSpanElement> {
  visuallyHiddenText?: false | string;
}

const ErrorMessage: React.FC<ErrorMessageProps> = ({
  className,
  visuallyHiddenText,
  children,
  ...rest
}) => (
  <span className={classNames('nhsuk-error-message', className)} {...rest}>
    {visuallyHiddenText !== false ? (
      <span className="nhsuk-u-visually-hidden">{visuallyHiddenText}</span>
    ) : null}
    {children}
  </span>
);

ErrorMessage.defaultProps = {
  visuallyHiddenText: 'Error: ',
  role: 'alert',
};

export default ErrorMessage;
