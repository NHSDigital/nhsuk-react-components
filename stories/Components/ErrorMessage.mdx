import {Meta} from "@storybook/blocks";

<Meta title="Components/ErrorMessage" />

# ErrorMessage

This component can be found in the `nhsuk-frontend` repository <a href="https://github.com/nhsuk/nhsuk-frontend/tree/master/packages/components/error-message" target="_blank" rel="noopener noreferrer">here</a>.

## Implementation Notes

The `ErrorMessage` component has a default `visuallyHiddenText` of "Error: ". This can be overriden, or disabled using `visuallyHiddenText={false}`.

## Usage

### Standard

```jsx
import { ErrorMessage } from "nhsuk-react-components";

const Element = () => {
    return (
        <ErrorMessage>Error!</ErrorMessage>
    );
}
```
