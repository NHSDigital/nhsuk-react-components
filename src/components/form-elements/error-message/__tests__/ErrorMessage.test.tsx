import React from 'react';
import { render } from '@testing-library/react';
import ErrorMessage from '../';

describe('ErrorMessage', () => {
  it('matches snapshot', () => {
    const { container } = render(<ErrorMessage>Error</ErrorMessage>);

    expect(container).toMatchSnapshot('ErrorMessage');
  });

  it('has default visuallyHiddenText', () => {
    const { container } = render(<ErrorMessage>Error</ErrorMessage>);

    expect(container.querySelector('.nhsuk-u-visually-hidden')?.textContent).toBe('Error: ');
  });

  it('has disabled visuallyHiddenText', () => {
    const { container } = render(<ErrorMessage visuallyHiddenText={false}>Error</ErrorMessage>);

    expect(container.querySelector('.nhsuk-u-visually-hidden')).toBeFalsy();
  });

  it('has custom visuallyHiddenText', () => {
    const { container } = render(<ErrorMessage visuallyHiddenText="Custom">Error</ErrorMessage>);

    expect(container.querySelector('.nhsuk-u-visually-hidden')?.textContent).toBe('Custom');
  });
});
