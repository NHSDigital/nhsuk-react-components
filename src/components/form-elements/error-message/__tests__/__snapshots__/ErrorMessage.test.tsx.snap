// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorMessage matches snapshot: ErrorMessage 1`] = `
<span
  className="nhsuk-error-message"
  role="alert"
>
  <span
    className="nhsuk-u-visually-hidden"
  >
    Error: 
  </span>
  Error
</span>
`;
